File: practice/mathematics/number_theory/eulers_criterion/eulers_criterion.hs
module Main where

import Control.Monad (forM_)

pow_mod_util :: Integer -> Integer -> Integer -> Integer -> Integer
pow_mod_util a 0 m res = res
pow_mod_util a deg m res
    | deg `mod` 2 == 1 = pow_mod_util (a * a `mod` m) (deg `div` 2) m (res * a `mod` m)
    | otherwise        = pow_mod_util (a * a `mod` m) (deg `div` 2) m res

pow_mod :: Integer -> Integer -> Integer -> Integer
pow_mod a d p = pow_mod_util a d p 1

euler :: Integer -> Integer -> Bool
euler a p
    | x == 1 = True
    | x /= 1 = False
    where
        d = (p - 1) `div` 2
        x = pow_mod a d p

main :: IO ()
main = do
    t <- readLn :: IO Int
    forM_ [1..t] $ \_ -> do
        ap <- getLine
        let [a, p] = map (read :: String -> Integer) $ words ap
        if p == 2 || a == 0 || euler a p == True
            then putStrLn "YES"
            else putStrLn "NO"

File: practice/mathematics/number_theory/eulers_criterion/eulers_criterion.cpp
#include <iostream>

using namespace std;

unsigned long long pow_mod(unsigned long long i_a, unsigned long long i_deg, unsigned long long i_mod)
{
    unsigned long long res = 1;
    
    while (i_deg > 0)
    {
        if (i_deg & 1 == 1)
        {
            res *= i_a;
            if (res >= i_mod)
            {
                res %= i_mod;
            }
        }
        
        i_a *= i_a;
        if (i_a >= i_mod)
        {
            i_a %= i_mod;
        }
        
        i_deg >>= 1;
    }
    
    return res;
}

bool euler(unsigned long long i_a, unsigned long long i_p)
{
    unsigned long long x = pow_mod(i_a, (i_p - 1) / 2, i_p);
    if (x == 1)
    {
        return true;
    }
    return false;
}

int main()
{
    int t;
    unsigned long long a, p;
    
    cin >> t;
    while (t--)
    {
        cin >> a >> p;
        if (p == 2 || a == 0 || euler(a, p))
        {
            cout << "YES" << endl;
        }
        else
        {
            cout << "NO" << endl;
        }
    }
    
    return 0;
}

